README.md
# naval 

Neat and Awesome Video Analytics Logging Library.

## Brief description

![demo](docs/demo.gif)

This library provides a logging-like interface to visualize debug information in OpenCV applications working with video streams, for example, video analytics.
In this case, one might what to visualize some debug information (like bounding boxes of detected objects) directly on the video frame.
This library provides a nice and easy way of achieving that.

## Features

- Logging arbitrary polygon-like primitives (rects, points, etc)
- Logging images (useful for visualizing preprocessed neural net inputs)
- Specifying log level (debug, info, etc) for each primitive for easy filtering
- Logging arbitrary metatada using tags (see demo above)

## Examples

### Basic usage

```cpp
#include <naval/log_packet.hpp>
#include <naval/sink.hpp>
#include <naval/stream_logger.hpp>
#include <opencv2/videoio.hpp>

int main() {
  cv::VideoCapture video_capture{"video.mp4"};
  naval::StreamLogger stream_logger{naval::CreateFileSink("video.mp4.navalil")};

  while (true) {
    cv::Mat frame;
    const double timestamp_sec = video_capture.get(cv::CAP_PROP_POS_MSEC) / 1000;
    if (!video_capture.read(frame)) {
      break;
    }

    naval::LogPacket log_packet{timestamp_sec};

    // Log stuff here

    stream_logger.WritePacket(log_packet);
  }

  return 0;
}
```

---

### Logging user-defined types

```cpp
namespace my_namespace {

// This is user-defined rect that that we want to log
struct MyRect {
  MyRect(float x, float y, float width, float height) : x{x}, y{y}, width{width}, height{height} {
  }

  float x;
  float y;
  float width;
  float height;
};

}  // namespace my_namespace

namespace naval {

// User has to provide this function to convert from his types into naval primitives
template <>
inline std::vector<Vertex> ConvertToVertices(const my_namespace::MyRect& geometry) {
  const auto [x, y, width, height] = geometry;
  return {{x + width, y}, {x + width, y + height}, {x, y + height}, {x, y}};
}

}  // namespace naval

void LogExample() {
  naval::LogPacket log_packet{0.0};

  my_namespace::MyRect rect{300, 300, 300, 300};
  log_packet.Info(rect, {{"Object type", "Simple rect"}, {"Metadata", "Lots of information!"}});
}
```

---

### Logging images
```cpp
#include <naval/log_packet.hpp>
#include <opencv2/core.hpp>

void LogImage(const cv::Mat& image) {
  LogPacket log_packet{0.0};

  log_packet.Info(image, {{"image_type", "nn_input_preprocessed"}});
}
```

---

## Requirements

- C++ 17 compiler
- CMake
- OpenCV>=4.5 (untested on older versions)

## Installation

- Simply use CMake `add_subdirectory`
- Alternatively, you can use `cpm-cmake`

## naval-player

`naval-player` is a simple PySide-based player for `.navalil` files. _Will be open-sourced later._

### Installation

1. [Download](https://disk.yandex.ru/d/vytI6d_4fjq3Tg) the .whl package.
2. On Ubuntu-based distros:
```bash
sudo apt-get install ubuntu-restricted-extras
```
3. Install the package:
```bash
pip install naval_player-0.4.5-py3-none-any.whl
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [v0.4.0](https://github.com/nvi-research/naval/releases/tag/v0.4.0) - 2023-09-06

<small>[Compare with v0.3.1](https://github.com/nvi-research/naval/compare/v0.3.1...v0.4.0)</small>

### Features

- popup messages ([3687351](https://github.com/nvi-research/naval/commit/3687351a697428641ba3e0889a224d792dd8559d) by Vyacheslav Zhdanovskiy).

## [v0.3.1](https://github.com/nvi-research/naval/releases/tag/v0.3.1) - 2023-04-07

<small>[Compare with v0.3.0](https://github.com/nvi-research/naval/compare/v0.3.0...v0.3.1)</small>

## [v0.3.0](https://github.com/nvi-research/naval/releases/tag/v0.3.0) - 2022-12-30

<small>[Compare with v0.2.0](https://github.com/nvi-research/naval/compare/v0.2.0...v0.3.0)</small>

### Features

- increase protocol version ([d8bba19](https://github.com/nvi-research/naval/commit/d8bba19276652fb3e7abe74d59102c4d43efe58b) by Vyacheslav Zhdanovskiy).
- store timestamp in LogPacket ([0f490ba](https://github.com/nvi-research/naval/commit/0f490bac6a16556beee8e822f359e56894fbdf6b) by Vyacheslav Zhdanovskiy).
- add log_level to Tag ([16bda62](https://github.com/nvi-research/naval/commit/16bda621b41d26dec10fd418eb632e701798e2de) by Vyacheslav Zhdanovskiy).
- remove NAVAL_DEBUG and NAVAL_INFO macros, restore Debug and Info methods ([52be9c4](https://github.com/nvi-research/naval/commit/52be9c4cd6d247224af3ff0731faa294aa5291a0) by Vyacheslav Zhdanovskiy).

## [v0.2.0](https://github.com/nvi-research/naval/releases/tag/v0.2.0) - 2022-12-29

<small>[Compare with v0.1.0](https://github.com/nvi-research/naval/compare/v0.1.0...v0.2.0)</small>

### Features

- increase protocol version ([3749d90](https://github.com/nvi-research/naval/commit/3749d90f777c23d2f9ed0ce49f237cfee3af194e) by Vyacheslav Zhdanovskiy).
- allow to track context in MessageMetadata ([085977a](https://github.com/nvi-research/naval/commit/085977abca782f8cc178ed4ee1a795571af336d7) by Vyacheslav Zhdanovskiy).
- new colors in palette ([8a0d8bc](https://github.com/nvi-research/naval/commit/8a0d8bc14f666f79cba55931dccce5234701a358) by Ilya Chaykin).

### Bug Fixes

- allow to use NAVAL_* macros outside of naval namespace ([d286483](https://github.com/nvi-research/naval/commit/d286483ad352d674afbcfbeb39acc2a1e683e42f) by Vyacheslav Zhdanovskiy).

### Code Refactoring

- refactor serialization code ([66d9a9d](https://github.com/nvi-research/naval/commit/66d9a9dadeec5556f2d5d8c529ed579af1f84a5b) by Vyacheslav Zhdanovskiy).

## [v0.1.0](https://github.com/nvi-research/naval/releases/tag/v0.1.0) - 2022-12-19

<small>[Compare with v0.0.1](https://github.com/nvi-research/naval/compare/v0.0.1...v0.1.0)</small>

### Features

- add more colors to palette ([887862c](https://github.com/nvi-research/naval/commit/887862cf2b81ab1ee0109c6fd8702ce56eaf0c2b) by Vyacheslav Zhdanovskiy).
- upgrade protocol version ([895602e](https://github.com/nvi-research/naval/commit/895602e6981bc588fea148db50f39e54a03fbad6) by Vyacheslav Zhdanovskiy).
- add line_thickness to DrawProperties ([d02c6a0](https://github.com/nvi-research/naval/commit/d02c6a0ba59f93de658ef53b0b8d2c365350da00) by Vyacheslav Zhdanovskiy).
- add font_size to TagProperties ([ac4391b](https://github.com/nvi-research/naval/commit/ac4391b44ac57958c8ee12b497d67a3537d81b13) by Vyacheslav Zhdanovskiy).

## [v0.0.1](https://github.com/nvi-research/naval/releases/tag/v0.0.1) - 2022-12-15

<small>[Compare with first commit](https://github.com/nvi-research/naval/compare/476539231851f77c20c63867df92dbfe464c8686...v0.0.1)</small>

### Features

- write stream_url to log ([234853a](https://github.com/nvi-research/naval/commit/234853a93b2c51ed104193fd01d13c1e84aab677) by Vyacheslav Zhdanovskiy).
- write file signature and protocol version to stream file ([c9516da](https://github.com/nvi-research/naval/commit/c9516da4b2ef3b335109831e1386052badb26fc6) by Vyacheslav Zhdanovskiy).
- architecture simplification ([8809e79](https://github.com/nvi-research/naval/commit/8809e79968a5ebbcadddbbe78437f95a63e7af46) by Vyacheslav Zhdanovskiy).
- basic working version ([e13cdd0](https://github.com/nvi-research/naval/commit/e13cdd0274ec2c6805b249d6675fee9f4858d4fd) by Vyacheslav Zhdanovskiy).

### Bug Fixes

- properly include sanitizers.cmake ([563dd3c](https://github.com/nvi-research/naval/commit/563dd3c7947b397d710bf88981977566a3e19257) by Vyacheslav Zhdanovskiy).
- lenna path in tests ([9841b67](https://github.com/nvi-research/naval/commit/9841b67908a9a3e95e0a7c3c2000923dcfeb7845) by Vyacheslav Zhdanovskiy).
- remove unused LogLevel ([678f420](https://github.com/nvi-research/naval/commit/678f420df638c234fbf10f7ceaffcf6bbef96132) by Vyacheslav Zhdanovskiy).

### Code Refactoring

- rename WriteFrame to WritePacket ([e7d938a](https://github.com/nvi-research/naval/commit/e7d938a500e2f5a2e38f6d0b92886d2c84075688) by Vyacheslav Zhdanovskiy).
- newlines in .clang-format and .gitignore ([28c93eb](https://github.com/nvi-research/naval/commit/28c93ebce2c15d68d8936490cda9401f472cd6f8) by Vyacheslav Zhdanovskiy).



examples/example.cpp
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2023, NVI Research
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <filesystem>
#include <iostream>
#include <vector>

#include <naval/log_packet.hpp>
#include <naval/sink.hpp>
#include <naval/stream_logger.hpp>
#include <opencv2/videoio.hpp>

namespace fs = std::filesystem;

namespace my_namespace {

// This is user-defined rect that that we want to log
struct MyRect {
  MyRect(float x, float y, float width, float height) : x{x}, y{y}, width{width}, height{height} {
  }

  float x;
  float y;
  float width;
  float height;
};

}  // namespace my_namespace

namespace naval {

// User has to provide this function to convert from his types into naval primitives
template <>
inline std::vector<Vertex> ConvertToVertices(const my_namespace::MyRect& geometry) {
  const auto [x, y, width, height] = geometry;
  return {{x + width, y}, {x + width, y + height}, {x, y + height}, {x, y}};
}

}  // namespace naval

int main() {
  // Video is located alongside the source .cpp
  const fs::path video_path = fs::path{__FILE__}.remove_filename() / "bunny.mp4";
  // Log file location is small_bunny_1080p_60fps.mp4.navalil
  const fs::path log_path = fs::path{video_path}.replace_extension(".mp4.navalil");
  cv::VideoCapture video_capture{video_path.string()};

  naval::StreamLogger stream_logger{naval::CreateFileSink(log_path)};

  while (true) {
    cv::Mat frame;
    const double timestamp_sec = video_capture.get(cv::CAP_PROP_POS_MSEC) / 1000;
    std::cout << "Processing frame with ts_sec = " << timestamp_sec << "\n";
    if (!video_capture.read(frame)) {
      break;
    }

    my_namespace::MyRect rect{300, 300, 300, 300};
    naval::LogPacket log_packet{timestamp_sec};
    log_packet.Info(rect, {{"Object type", "Simple rect"}, {"Metadata", "Lots of information!"}});

    stream_logger.WritePacket(log_packet);
  }

  return 0;
}
